tmt/steps/prepare/artifact/providers/koji.py

```diff
@@ -184,3 +234,136 @@ def _download_artifact(
+        version = rpm_meta["version"]
+        release = rpm_meta["release"]
+        arch = rpm_meta["arch"]
+
+        # Construct the full URL for this RPM
+        url = (
+            f"{self._topurl}/packages/{name}/"
+            f"{version}/{release}/{arch}/"
+            f"{name}-{version}-{release}.{arch}.rpm"
+        )
+
+        return RpmArtifactInfo(_raw_artifact={**rpm_meta, "url": url})
+
+    def make_scratch_artifact(self, task_id: int, filename: str) -> KojiScratchRpmArtifactInfo:
+        """
+        Create a scratch RPM artifact from a task output filename.
+        """
+        pathinfo = koji.PathInfo(  # type: ignore[union-attr]
+            topdir=self._topurl
+        )
+        work_path = pathinfo.work("DEFAULT")
+        task_path = pathinfo.taskrelpath(task_id)
+        url = f"{work_path}/{task_path}/{filename}"
+
+        raw_artifact = {
+            "filename": filename,
+            "url": url,
+        }
+        return KojiScratchRpmArtifactInfo(_raw_artifact=raw_artifact)
+
+
+@provides_artifact_provider("koji.task")  # type: ignore[arg-type]
+class KojiTask(KojiArtifactProvider):
+    @cached_property
+    def build_id(self) -> Optional[int]:
+        task_id = int(self.id)
+        if builds := self._call_api("listBuilds", taskID=task_id):
+            if len(builds) > 1:
+                self.logger.warning(
+                    f"Task '{task_id}' produced {len(builds)} builds, using the first one."
+                )
+            build_id = builds[0]["build_id"]  # Assume the task produced a single build
+            assert isinstance(build_id, int)
+            return build_id
+        return None
+
+    def _get_task_children(self, task_id: int) -> list[int]:
+        """
+        Recursively fetch all child tasks of the given task ID.
+
+        :param task_id: The parent task ID
+        :return: List of all child task IDs
+        """
+        child_tasks: list[int] = [task_id]  # Include the parent task itself
+        direct_children = self._call_api("getTaskChildren", task_id)
+        for child in direct_children:
+            child_id = child["id"]
+            assert isinstance(child_id, int)
+            child_tasks.append(child_id)
+            # Recursively fetch grandchildren
+            child_tasks.extend(self._get_task_children(child_id))
+        return child_tasks
+
+    @cached_property
+    def rpm_list(self) -> list[RpmArtifactInfo]:
+        self.logger.debug(f"Fetching RPMs for task '{self.id}'.")
+        # If task produced a build, reuse build path
+        if self.build_id is not None:
+            self.logger.debug(
+                f"Task '{self.id}' produced build '{self.build_id}', fetching RPMs from the build."
+            )
+            return self._get_build_provider(self.build_id).rpm_list
+
+        # Otherwise, list the task output files for scratch builds
+        self.logger.debug(f"Task '{self.id}' did not produce a build, fetching scratch RPMs.")
```

Comment:
I would rather we reverse the log, for me it is more unusual that we would pass the taskID if we can also know the buildID, and that would be more interesting to report to see if they actually need to do that.

Reply:
Yeah but verbosity and messaging. But we can get to that later